airflow/decorators/base.py

```diff
@@ -210,17 +212,27 @@ def __init__(
         super().__init__(task_id=task_id, **kwargs_to_upstream, **kwargs)
 
     def execute(self, context: Context):
+        for arg in chain(self.op_args, self.op_kwargs.values()):
+            if isinstance(arg, Dataset):
+                self.inlets.append(arg)
```

Comment:
Shouldn't this be done in the prepare_lineage function (which I think will already called by `super().execute(context)` a few lines down.

Reply:
Are you suggesting that we would make this more generic? It now only affects the `DecoratedOperator` moving it to prepare_lineage (part of `pre_execute`) would make it affect all.

Reply:
That was my thought (specifically why should taskflow be treated specially?) but I haven't thought that deeply to consider what the implications might be.

Reply:
I don't think taskflow should be treated separately. It was my starting point (i.e. let taskflow support datasets). I can have a look at what it takes to move it a level higher.